ddtrace/tracer/sampler.go

```diff
@@ -147,3 +154,249 @@ func (ps *prioritySampler) apply(spn *span) {
+			return nil
+		}
+		for _, v := range jsonRules {
+			if v.Rate == "" {
+				log.Warn("error parsing rule: rate not provided")
+				continue
+			}
+			rate, err := v.Rate.Float64()
+			if err != nil {
+				log.Warn("error parsing rule: invalid rate: %v", err)
+				continue
+			}
+			switch {
+			case v.Service != "" && v.Operation != "":
+				rules = append(rules, ServiceOperationRule(v.Service, v.Operation, rate))
+			case v.Service != "":
+				rules = append(rules, ServiceRule(v.Service, rate))
+			case v.Operation != "":
+				rules = append(rules, OperationRule(v.Operation, rate))
+			}
+		}
+	}
+	validRules := make([]SamplingRule, 0, len(rules))
+	for _, v := range rules {
+		if !(v.Rate >= 0.0 && v.Rate <= 1.0) {
+			log.Warn("ignoring rule %+v: rate is out of range", v)
+			continue
+		}
+		validRules = append(validRules, v)
+	}
+	return validRules
+}
+
+// sampleRate returns the sampling rate found in the DD_TRACE_SAMPLE_RATE environment variable. If it is invalid and
+// not within the 0-1 range, the default value of 0 is returned.
+func sampleRate() float64 {
+	const defaultRate = 0.0
+	v := os.Getenv("DD_TRACE_SAMPLE_RATE")
+	if v == "" {
+		return defaultRate
+	}
+	r, err := strconv.ParseFloat(v, 64)
+	if err != nil {
+		log.Warn("using default rate %f because DD_TRACE_SAMPLE_RATE is invalid: %v", defaultRate, err)
+		return defaultRate
+	}
+	if r >= 0.0 && r <= 1.0 {
+		return r
+	}
+	log.Warn("using default rate %f because provided value is out of range: %f", defaultRate, r)
+	return defaultRate
+}
+
+// newRateLimiter returns a new rate limiter which allows r events per second.
+// The DD_TRACE_LIMIT environment variable may override r.
+func newRateLimiter(r float64) *rate.Limiter {
+	defaultLimiter := rate.NewLimiter(rate.Inf, 0)
+	v := os.Getenv("DD_TRACE_RATE_LIMIT")
+	if v == "" {
+		return defaultLimiter
+	}
+	l, err := strconv.ParseFloat(v, 64)
+	if err != nil {
+		log.Warn("using default rate limit because DD_TRACE_RATE_LIMIT is invalid: %v", err)
+		return defaultLimiter
+	}
+	if l < 0.0 {
+		log.Warn("using default rate limit because DD_TRACE_RATE_LIMIT is negative: %f", l)
+		return defaultLimiter
+	}
+	return rate.NewLimiter(rate.Limit(l), int(math.Ceil(r*l)))
+}
+
+// apply uses the sampling rules to determine the sampling rate for the
+// provided span. If the rules don't match, and a default rate hasn't been
+// set using DD_TRACE_SAMPLE_RATE, then it returns false and the span is not
+// modified.
+func (rs *rulesSampler) apply(span *span) bool {
+	var matched bool
+	rate := rs.rate
+	for _, rule := range rs.rules {
+		if rule.match(span) {
+			matched = true
+			rate = rule.Rate
+			break
+		}
+	}
+	if !matched && rate == 0.0 {
+		// no matching rule or global rate, so we want to fall back
+		// to priority sampling
+		return false
+	}
+
+	rs.applyRate(span, rate, time.Now())
+	return true
+}
+
+func (rs *rulesSampler) applyRate(span *span, rate float64, ts time.Time) {
```

Comment:
```suggestion
func (rs *rulesSampler) applyRate(span *span, rate float64, now time.Time) {
```
Can we please call it `now`? Seems more friendly and common. If you don't like `now`, maybe `t`?